### PR TITLE
Refactor VisionTECudaGraphHelper to minimize overrides and clarify state tracking

### PR DESCRIPTION
**Pull request by buptzyb**
# What does this PR do ?
## Summary

This pull request refactors the CUDA Graph helper class hierarchy based on review feedback from #3293. Key improvements include:

1. **Barrier-free synchronization**: Replaced `torch.distributed.barrier()` with `torch.cuda.synchronize()` to eliminate deadlock risks
2. **Minimized method overrides**: Reduced VisionTECudaGraphHelper from 3 overrides (38 lines) to 2 minimal overrides (8 lines)
3. **Extracted cleanup logic**: New `_reset_after_capture()` method allows clean no-op override in vision subclass
4. **Two-flag design**: Separated "capture process ran" from "graphs exist" with intuitive naming
5. **Updated training loop**: Uses appropriate APIs for state tracking vs. resource management

### Parent Class (TECudaGraphHelper)

**Synchronization Changes**
- **`_start_capturing()`**: Replaced `torch.distributed.barrier()` with `torch.cuda.synchronize()`
  - Eliminates deadlock risk when not all ranks call `create_cudagraphs()`
  - Each rank independently synchronizes its own CUDA operations before capture
  - No cross-rank coordination needed
- **`_finish_capturing()`**: Replaced `torch.distributed.barrier()` with `torch.cuda.synchronize()`
  - Ensures local CUDA operations complete before cleanup
  - Allows ranks to finish capture independently

**Structural Changes**
- Added `_reset_after_capture()`: Extracts model/optimizer reset logic into overridable method
- Updated `create_cudagraphs()`: Handles empty `flattened_callables` case; all ranks now call `_start_capturing()` and `_finish_capturing()` for symmetric execution
- Two-flag state tracking:
  - `_capture_ran`: Whether `create_cudagraphs()` completed (for training loop)
  - `_graphs_created`: Whether actual graphs exist (for resource management)
- New public methods:
  - `capture_ran()`: Returns `_capture_ran`
  - `graphs_created()`: Returns `_graphs_created` (renamed semantics!)

### Vision Subclass (VisionTECudaGraphHelper)

- **Removed overrides** (no longer needed):
  - `create_cudagraphs()` (21 lines) - parent handles empty layers case
  - `_start_capturing()` (17 lines) - CUDA sync eliminates need for custom barrier handling
- **Added clean overrides**:
  - `_reset_after_capture()`: No-op (vision cleanup handled by LM helper)
  - `_finish_capturing()`: Wraps graphs + calls `super()` (no duplication)
- **Updated docstring**:
  - Added note that ranks without vision layers can safely skip calling `create_cudagraphs()`
  - Clarified independent operation with PP > 1

### Training Loop (training.py)

Updated to use semantically correct APIs:
- `capture_ran()`: Process state checks (before capture, after warmup)
- `graphs_created()`: Resource checks (before deletion)

### Barrier-Free Architecture
- **No deadlock risk**: Ranks can independently call `create_cudagraphs()` without coordination
- **Better performance**: No network communication overhead from distributed barriers
- **Correct semantics**: We only need to ensure local GPU work is done, not cross-rank coordination
- **Flexible usage**: With PP > 1, only ranks with graphable layers need to call `create_cudagraphs()`

### Code Quality
- **Cleaner architecture**: Minimal, focused overrides in subclass
- **Symmetric execution**: All ranks follow same path when they do call `create_cudagraphs()`
- **Better maintainability**: No duplicated cleanup logic
- **Intuitive semantics**: `graphs_created()` means what it says
- **Clear intent**: Separate APIs for process state vs. resource existence

## Testing

- **Updated unit tests** to use new two-flag API:
  - Added `capture_ran()` assertions to verify both flags
  - Tests verify correct behavior when no graphable layers found:
    - `capture_ran()` returns `True` (process completed)
    - `graphs_created()` returns `False` (no graphs to delete)
- **PP=2 tests** validate independent operation:
  - Stage 0 creates graphs for vision layers
  - Stage 1 gracefully handles absence of vision layers
  - No deadlock when only stage 0 calls `create_cudagraphs()`

### Why CUDA Sync Instead of Distributed Barrier?

| Aspect | `torch.distributed.barrier()` | `torch.cuda.synchronize()` |
|--------|------------------------------|---------------------------|
| **Scope** | All ranks in process group | Current GPU device only |
| **Deadlock risk** | Yes - if some ranks don't call it | No - each rank independent |
| **Purpose** | Cross-rank coordination | Ensure GPU work completed |
| **Use case** | Coordinating distributed state | Preparing for CUDA Graph capture |
| **Performance** | Network communication overhead | Local synchronization only |

**Before capture**: We need to ensure all prior CUDA operations on the current device have completed so the graph capture doesn't race with ongoing GPU work. This is a **local synchronization requirement**, not a cross-rank coordination requirement.

**After capture**: We need to ensure the graph has been fully captured before we start cleanup operations. Again, this is **local to each rank**.

### Edge Cases Handled

1. **PP > 1 with vision on first stage only**:
   - Stage 0 creates graphs independently
   - Stage 1 can skip calling `create_cudagraphs()` entirely (no deadlock)
   - Or call it and gracefully return with no graphs created

2. **No graphable layers found**:
   - `capture_ran()` returns `True` (process ran)
   - `graphs_created()` returns `False` (no cleanup needed)

3. **Mixed calling patterns**:
   - Some ranks can create graphs while others don't
   - No synchronization required between them

## Related

Addresses review feedback from @buptzyb on #3293:
https://github.com/NVIDIA/Megatron-LM/pull/3293#pullrequestreview-3874229620

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by lhb8125** [comment]
/ok to test [1f409c5](https://github.com/NVIDIA/Megatron-LM/commit/1f409c5a12f243e1337f54e85fffa5a4426e28db)

**Comment by lhb8125** [comment]
/ok to test [ef074b2](https://github.com/NVIDIA/Megatron-LM/pull/3748/commits/ef074b27b3a0aac024e9904fbd66b7ca8e506158)

File: megatron/core/transformer/cuda_graphs.py [hunk 1]
```diff
@@ -2651,45 +2684,27 @@ def _discover_layers(self):
...
-            f'{torch.distributed.get_rank()}: {time.time() - start_time}s',
-        )
-        _set_capture_end()
-        if FREEZE_GC:
-            gc.unfreeze()
-        gc.collect()
-        torch.cuda.empty_cache()
-        self._graphs_created = True
+
+    def _finish_capturing(self, start_time):
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/cuda_graphs.py:2701]: I see, you worried that not all ranks call `create_cudagraphs()`.  How about we no longer use the barrier, but a synchronization instead?

**Review by jiemingz — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

File: megatron/training/training.py [hunk 2]
```diff
@@ -2836,7 +2836,7 @@ def trace_handler(p):
         # Capture CUDA Graphs.
         if (
             args.cuda_graph_impl == "transformer_engine"
-            and not cuda_graph_helper.graphs_created()
+            and not cuda_graph_helper.capture_ran()
```
Comment (ericharper) [inline megatron/training/training.py]: Does bridge need this updated as well? @maanug-nv , @gautham-kollu  ?

**Review by ericharper — APPROVED** [review]

File: megatron/training/training.py (hunk 2 above)
Comment (gautham-kollu) [reply megatron/training/training.py]: Yes.

File: megatron/core/transformer/cuda_graphs.py [hunk 3]
```diff
@@ -1718,8 +1718,11 @@ def __init__(self, model, config, seq_length, micro_batch_size, optimizers=[]):
 
         self._discover_layers()
 
-        # One helper object can only capture CUDA Graphs once. Use this flag to check if the graphs
-        # have been created.
+        # Flags to track CUDA Graph state:
+        # - _capture_ran: Whether create_cudagraphs() has been called (used by training loop)
+        # - _graphs_created: Whether any graphs were actually created (may be False if no
+        #   layers found)
+        self._capture_ran = False
```
Comment (gautham-kollu) [inline megatron/core/transformer/cuda_graphs.py]: Can we rename this to "`_capture_finished`" ? kinda borrowing from "_graph_created" shows the capture was finished ?

Comment (buptzyb) [reply megatron/core/transformer/cuda_graphs.py]: done

**Review by maanug-nv — APPROVED** [review]

File: megatron/training/training.py (hunk 2 above)
Comment (maanug-nv) [reply megatron/training/training.py]: yes. until we finish upstreaming, syncing is supposed to be part of oncall's responsibilities.
